### PR TITLE
Test dynamic field conservation rules

### PR DESCRIPTION
**Pull request by dariorussi**
## Description 

Test dynamic field conservation rules

## Test Plan 

existing tests

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by sblackshear** [comment]
Yep that is the one

File: crates/sui-types/src/temporary_store.rs [hunk 1]
```diff
@@ -660,10 +660,10 @@ impl<S: GetModule + ObjectStore + BackingPackageStore> TemporaryStore<S> {
     /// This intended to be called *after* we have charged for gas + applied the storage rebate to the gas object,
     /// but *before* we have updated object versions
     pub fn check_sui_conserved(&self) {
-        if !self.dynamic_fields_touched().is_empty() {
-            // TODO: check conservation in the presence of dynamic fields
-            return;
-        }
+        // if !self.dynamic_fields_touched().is_empty() {
```
Comment (tnowacki) [inline crates/sui-types/src/temporary_store.rs:663]: I think this function needs to be fixed. I _think_ you can more or less remove this function and just check all writes/deletes? I'm not really sure why the two are separated

Comment (sblackshear) [reply crates/sui-types/src/temporary_store.rs:663]: The problem here is that we need to see how much SUI was in the dynamic field before the write/delete. For that, we need the version of the object at input, and that info is not present in `writes` or `deletes`

Comment (tnowacki) [reply crates/sui-types/src/temporary_store.rs:663]: For writes you have `obj.version` and for deletes, there is the version as a part of the delete information. Unless these aren't the writes/deletes I think they are

Comment (sblackshear) [reply crates/sui-types/src/temporary_store.rs:663]: Ah, maybe I misunderstand:
- For `deletes`, you have a `SequenceNumber` that is presumably the input version https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/temporary_store.rs#L46
- For `writes`, you have an `ObjectRef` that I was assuming is the output version (let me know if I'm wrong about that) https://github.com/MystenLabs/sui/blob/main/crates/sui-types/src/temporary_store.rs#L45. If that's right, it's easy to find the input version for objects that are explicit inputs (as the current logic does), but not for dynamic fields (which the current logic is trying to do, but maybe incorrectly).

Comment (tnowacki) [reply crates/sui-types/src/temporary_store.rs:663]: It should not be the output version. I think the new version is only set at `into_effects` (or `to_effects`). I'm basing this on the logic of `modified_at_versions` which is the "input" version.

Comment (sblackshear) [reply crates/sui-types/src/temporary_store.rs:663]: Ah sorry, the code I linked to was `InnerTemporaryStore`, not `TemporaryStore` (which this function is for). But I think I see how to do this without `dynamic_fields_touched` if we move the function to `InnerTemporaryStore`.

Comment (tnowacki) [reply crates/sui-types/src/temporary_store.rs:663]: Er bit confused. But for clarity, the properties for written/deletes I was talking about are for the `TemporaryStore`, not the `InnerTemporaryStore`. Which do you need to write the function over?

Comment (sblackshear) [reply crates/sui-types/src/temporary_store.rs:663]: I currently have it written over `TemporaryStore` and would prefer to keep it there.

**Pull request closed without merging**